cli/tools/test/channel.rs

```diff
@@ -0,0 +1,481 @@
+    let sender = self.sender.clone();
+
+    // Each worker spawns its own output monitoring and serialization task. This task will
+    // poll the stdout/stderr streams and interleave that data with `TestEvents` generated
+    // by the test runner worker.
+    //
+    // Note that this _must_ be a separate thread! Flushing requires locking coördination
+    // on two threads and if we're blocking-locked on the mutex we've sent down the sync_receiver,
+    // there's no way for us to process the actual flush operation here.
+    //
+    // Creating a mini-runtime to flush the stdout/stderr is the easiest way to do this, but
+    // there's no reason we couldn't do it with non-blocking I/O, other than the difficulty
+    // of setting up an I/O reactor in Windows.
+    std::thread::spawn(move || {
+      let runtime = tokio::runtime::Builder::new_current_thread()
+        .enable_io()
+        .build()
+        .unwrap();
+      runtime.block_on(tokio::task::unconstrained(async move {
+        let mut test_stdout = TestStream::new(
+          id,
+          TestStdioStream::Stdout,
+          stdout_reader,
+          sender.clone(),
+        )?;
+        let mut test_stderr =
+          TestStream::new(id, TestStdioStream::Stderr, stderr_reader, sender)?;
+
+        // This function will be woken whenever a stream or the receiver is ready
+        loop {
+          tokio::select! {
+            _ = test_stdout.pipe(), if test_stdout.is_alive() => {},
+            _ = test_stderr.pipe(), if test_stdout.is_alive() => {},
+            recv = sync_receiver.recv() => {
+              match recv {
+                // If the channel closed, we assume that all important data from the streams was synced,
+                // so we just end this task immediately.
+                None => { break },
+                Some(mutex) => {
+                  // If we fail to write the sync marker for flush (likely in the case where the runtime is shutting down),
+                  // we instead just release the mutex and bail.
+                  let success = stdout_writer.write_all(SYNC_MARKER).is_ok()
+                    && stderr_writer.write_all(SYNC_MARKER).is_ok();
+                  if success {
+                    for stream in [&mut test_stdout, &mut test_stderr] {
+                      stream.read_until_sync_marker().await;
+                    }
+                  }
+                  drop(mutex);
+                }
+              }
+            }
+          }
+        }
+
+        Ok::<_, std::io::Error>(())
+      }))?;
+
+      Ok::<_, std::io::Error>(())
+    });
+
+    let sender = TestEventSender {
+      id,
+      ref_count: Default::default(),
+      sender: self.sender.clone(),
+      sync_sender,
+    };
+
+    TestEventWorkerSender {
+      sender,
+      stdout,
+      stderr,
+    }
+  }
+
+  /// A [`TestEventWeakSender`] has a unique ID, but will not keep the [`TestEventReceiver`] alive.
```

Comment:
Can you update this comment and give an example usage of the `weak_sender`?

Reply:
Done.